digdag-standards/src/main/java/io/digdag/standards/operator/pg/PgConnection.java

```diff
@@ -0,0 +1,188 @@
+        throws NotReadOnlyException
+    {
+        try {
+            execute("SET TRANSACTION READ ONLY");
+            try (Statement stmt = connection.createStatement()) {
+                ResultSet rs = stmt.executeQuery(sql);  // executeQuery throws exception if given query includes multiple statements
+                resultHandler.accept(new PgResultSet(rs));
+            }
+            execute("SET TRANSACTION READ WRITE");
+        }
+        catch (SQLException ex) {
+            if (ex.getSQLState().equals("25006")) {  // 25006 = read_only_sql_transaction error
+                throw new NotReadOnlyException(ex);
+            }
+            else {
+                throw new DatabaseException("Failed to execute given SELECT statement", ex);
+            }
+        }
+    }
+
+    @Override
+    public String escapeIdent(String ident)
+    {
+        try {
+            StringBuilder buf = new StringBuilder();
+            escapeIdentifier(buf, ident);
+            return buf.toString();
+        }
+        catch (SQLException e) {
+            throw new IllegalArgumentException("Invalid identifier name (contains \\0 character): " + ident);
+        }
+    }
+
+    @Override
+    public TransactionHelper getStrictTransactionHelper(String statusTableName)
+    {
+        return new PgPersistentTransactionHelper(statusTableName);
+    }
+
+    private class PgPersistentTransactionHelper
+            extends AbstractPersistentTransactionHelper
+    {
+        PgPersistentTransactionHelper(String statusTableName)
+        {
+            super(statusTableName);
+        }
+
+        @Override
+        public void prepare()
+        {
+            executeStatement("create a status table " + escapeIdent(statusTableName),
+                    String.format(ENGLISH,
+                        "CREATE TABLE IF NOT EXISTS %s" +
+                        " (query_id text NOT NULL UNIQUE, created_at timestamptz NOT NULL, completed_at timestamptz)",
+                        escapeIdent(statusTableName))
+                   );
+        }
+
+        @Override
+        public void cleanup()
+        {
+            executeStatement("delete old query status rows from " + escapeIdent(statusTableName) + " table",
+                    String.format(ENGLISH,
+                        "DELETE FROM %s WHERE query_id = ANY(" +
+                        "SELECT query_id FROM %s WHERE completed_at < now() - interval '24' hours" +
+                        ")",
+                        escapeIdent(statusTableName),
+                        escapeIdent(statusTableName))
+                    );
+        }
+
+        @Override
+        protected StatusRow lockStatusRow(UUID queryId)
+        {
+            try (Statement stmt = connection.createStatement()) {
+                ResultSet rs = stmt.executeQuery(String.format(ENGLISH,
+                            "SELECT completed_at FROM %s WHERE query_id = '%s' FOR UPDATE",
```

Comment:
oh I see. yes, you're right.
If the target PostgreSQL supports SKIP LOCKED (PostgreSQL >= 9.5), here can be non-blocking so that this agent thread can run other operators meanwhile.